xrpl/clients/websocket_client.py

```diff
@@ -0,0 +1,119 @@
+
+    def request(self: WebsocketClient, request_object: Request) -> Response:
+        """
+        Synchronously submit the request represented by the request_object to the
+        rippled node specified by this client's URL.
+
+        Arguments:
+            request_object: An object representing information about a rippled request.
+
+        Returns:
+            The response from the server, as a Response object.
+        """
+        return asyncio.get_event_loop().run_until_complete(
+            self.request_async(request_object)
+        )
+
+    async def listen_async(
+        self: WebsocketClient,
+        request_object: Request,
+        handler: Callable[[Response], None],
+    ) -> None:
+        """
+        Asynchronously submits a request represented by the request_object to the
+        rippled node specified by this client's URL, and listens to (and processes) all
+        responses from the server.
+
+        Arguments:
+            request_object: An object representing information about a rippled request.
+            handler: The method that handles the Response objects from the server.
+                Takes a Response object as a parameter and returns None.
+        """
+        formatted_request = request_to_websocket(request_object)
+        if "id" not in formatted_request:
+            formatted_request["id"] = "request_{}".format(formatted_request["command"])
+
+        async with websockets.connect(self.url) as websocket:
+            await websocket.send(json.dumps(formatted_request))
+            try:
+                async for message in websocket:
+                    await websocket.send(json.dumps(formatted_request))
+                    response = await websocket.recv()
+                    response_dict = json.loads(response)
+                    assert response_dict["id"] == formatted_request["id"]
+                    handler(websocket_to_response(response_dict))
+            except asyncio.exceptions.CancelledError:
+                return
+
+    def listen(
+        self: WebsocketClient,
+        request_object: Request,
+        handler: Callable[[Response], None],
+    ) -> None:
+        """
+        Synchronously submits a request represented by the request_object to the
+        rippled node specified by this client's URL, and listens to (and processes) all
+        responses from the server.
+
+        Arguments:
+            request_object: An object representing information about a rippled request.
+            handler: The method that handles the Response objects from the server.
+                Takes a Response object as a parameter and returns None.
+        """
+        asyncio.get_event_loop().run_until_complete(
+            self.listen_async(request_object, handler)
+        )
+        # asyncio.get_event_loop().run_forever()
+
+    def close(self: WebsocketClient) -> None:
+        """Closes any open WebSocket connections."""
+        for task in asyncio.all_tasks():
```

Comment:
pretty sure this will cancel _all_ tasks regardless of whether they are relating to sockets. that's not right.

Reply:
you may need to store open requests related to the object, on the object to solve this